File: code/R/04_psm_dd.R
# ============================================================
#  04_psm_dd.R — Estimation PSM-DD (Heckman et al. 1997/1998)
#  ATT sur pauvre_AF et pauvre_MODA
# ============================================================

source("code/R/config.R")
source("code/R/utils.R")

enfants_2018 <- readRDS(file.path(OUTPUT_DIR, "enfants_dep_2018.rds"))
enfants_2021 <- readRDS(file.path(OUTPUT_DIR, "enfants_dep_2021.rds"))
traitement_2018 <- readRDS(file.path(OUTPUT_DIR, "traitement_2018.rds"))
traitement_2021 <- readRDS(file.path(OUTPUT_DIR, "traitement_2021.rds"))

# Welfare : pcexp, hhsize, region, milieu + caracteristiques du CM
wel_2018 <- lire_stata(BASE_2018, "ehcvm_welfare_sen2018.dta")
wel_2021 <- lire_stata(BASE_2021, "ehcvm_welfare_sen2021.dta")

ID <- c("grappe", "menage")

# ── Variables welfare selectionnees ──────────────────────────
# hgender (sexe CM), hage (age CM), heduc (education CM), hmstat (situation famille CM)
# region, milieu, hhsize, pcexp

VARS_WEL <- c(ID, "pcexp", "hhsize", "region", "milieu",
              "hgender", "hage", "heduc", "hmstat")

sel_wel <- function(wel) {
  vars_dispo <- intersect(VARS_WEL, names(wel))
  wel |>
    dplyr::select(dplyr::all_of(vars_dispo)) |>
    dplyr::mutate(dplyr::across(where(haven::is.labelled), haven::zap_labels))
}

# ── Fusion base analytique ────────────────────────────────────

construire_base <- function(enfants, traitement, welfare, annee, t_val) {
  enfants |>
    dplyr::mutate(dplyr::across(where(haven::is.labelled), haven::zap_labels)) |>
    dplyr::left_join(traitement |>
                       dplyr::mutate(dplyr::across(where(haven::is.labelled),
                                                   haven::zap_labels)) |>
                       dplyr::select(dplyr::all_of(c(ID, "D"))),
                     by = ID) |>
    dplyr::left_join(sel_wel(welfare), by = ID) |>
    dplyr::mutate(
      annee     = annee,
      t         = t_val,
      log_pcexp = log(pcexp + 1),
      # Facteurs pour le probit
      f_milieu  = as.factor(dplyr::coalesce(milieu, 1L)),
      f_region  = as.factor(dplyr::coalesce(region, 1L)),
      f_heduc   = as.factor(dplyr::coalesce(heduc,  0L)),
      f_hmstat  = as.factor(dplyr::coalesce(hmstat, 1L)),
      hgender_n = dplyr::coalesce(as.integer(hgender), 1L),
      hage_n    = dplyr::coalesce(as.numeric(hage), median(as.numeric(hage), na.rm=TRUE)),
      hhsize_n  = dplyr::coalesce(as.numeric(hhsize), median(as.numeric(hhsize), na.rm=TRUE))
    )
}

base_2018 <- construire_base(enfants_2018, traitement_2018, wel_2018, 2018, 0)
base_2021 <- construire_base(enfants_2021, traitement_2021, wel_2021, 2021, 1)

pseudo_panel <- dplyr::bind_rows(
  dplyr::mutate(base_2018, dplyr::across(where(haven::is.labelled), haven::zap_labels)),
  dplyr::mutate(base_2021, dplyr::across(where(haven::is.labelled), haven::zap_labels))
)

cat("Base analytique :", nrow(pseudo_panel),
    "obs | traites :", sum(pseudo_panel$D == 1, na.rm = TRUE), "\n")

# ── Statistiques descriptives par statut traitement ───────────

pseudo_panel |>
  dplyr::group_by(annee, D) |>
  dplyr::summarise(
    n          = dplyr::n(),
    pct_AF     = round(taux(pauvre_AF) * 100, 1),
    pct_MODA   = round(taux(pauvre_MODA) * 100, 1),
    score_moy  = round(taux(score_dep), 3),
    dep_moda   = round(taux(nb_dep), 2),
    pcexp_moy  = round(taux(pcexp), 0),
    .groups    = "drop"
  ) |>
  print()

# ── PSM : probit sur covariables (periode de base t=0) ───────
# Specification complete : caracteristiques du CM + composition + localisation

formule_probit <- D ~ hhsize_n + log_pcexp + f_milieu + f_region +
                      hgender_n + hage_n + f_heduc + f_hmstat

base_t0 <- base_2018 |>
  dplyr::filter(!is.na(D), !is.na(log_pcexp), !is.na(hhsize_n))

probit_mod <- glm(formule_probit, data = base_t0,
                  family = binomial(link = "probit"))
cat("\n=== Modele probit (score de propension) ===\n")
print(summary(probit_mod))

# Pseudo-R2 de McFadden
ll_null <- probit_mod$null.deviance / (-2)
ll_mod  <- probit_mod$deviance / (-2)
cat(sprintf("Pseudo-R2 McFadden : %.3f\n", 1 - ll_mod / ll_null))

base_t0 <- base_t0 |>
  dplyr::mutate(pscore = predict(probit_mod, type = "response"))

# Graphique overlap (support commun)
p_overlap <- ggplot2::ggplot(base_t0, ggplot2::aes(x = pscore, fill = factor(D))) +
  ggplot2::geom_density(alpha = 0.5) +
  ggplot2::scale_fill_manual(values = c("steelblue", "tomato"),
                              labels = c("Non-traites", "Traites")) +
  ggplot2::labs(title = "Distribution du score de propension — support commun",
                x = "Score de propension", y = "Densite", fill = "") +
  ggplot2::theme_minimal(base_size = 12)
ggplot2::ggsave(file.path(OUTPUT_DIR, "overlap.pdf"), p_overlap, width = 8, height = 5)

# ── Appariement 1 : k-NN (k=4, sans remplacement) ────────────

match_knn <- MatchIt::matchit(formule_probit, data = base_t0,
                               method = "nearest",
                               distance = "glm", link = "probit",
                               ratio = 4, replace = FALSE)
cat("\n=== Bilan appariement k-NN ===\n")
print(summary(match_knn, un = FALSE))

# Balance plot
cobalt::love.plot(match_knn,
                  threshold   = 0.1,
                  title       = "Balance avant/apres appariement (k-NN)",
                  var.order   = "unadjusted",
                  colors      = c("steelblue", "tomato"))

# ── Appariement 2 : Kernel (Epanechnikov) ────────────────────

match_kernel <- MatchIt::matchit(formule_probit, data = base_t0,
                                  method  = "full",
                                  distance = "glm", link = "probit",
                                  estimand = "ATT")

# ── Appariement 3 : Caliper (epsilon = 0.05) ─────────────────

match_cal <- MatchIt::matchit(formule_probit, data = base_t0,
                               method   = "nearest",
                               distance = "glm", link = "probit",
                               caliper  = 0.05, std.caliper = FALSE,
                               ratio = 1)

# ── Base appariee pour la DD ──────────────────────────────────

matched_data_knn <- MatchIt::match.data(match_knn)

# Reconstituer le pseudo-panel sur la base appariee
# (traites kNN + temoins correspondants pour t=0 et t=1)
id_traites <- matched_data_knn |>
  dplyr::filter(D == 1) |>
  dplyr::select(dplyr::all_of(ID)) |>
  dplyr::distinct()

id_temoins <- matched_data_knn |>
  dplyr::filter(D == 0) |>
  dplyr::select(dplyr::all_of(ID)) |>
  dplyr::distinct()

panel_apparie <- pseudo_panel |>
  dplyr::semi_join(
    dplyr::bind_rows(id_traites, id_temoins),
    by = ID
  )

cat("Panel apparie :", nrow(panel_apparie), "obs\n")

# ── Double Difference (DD simple, sans appariement) ──────────
# Y_it = alpha + beta*t + gamma*D + delta*(t x D) + eps
# delta = ATT_DD

cat("\n=== Double Difference (sans appariement) ===\n")
for (outcome in c("pauvre_AF", "pauvre_MODA")) {
  formule_dd <- as.formula(paste(outcome, "~ factor(t) + D + factor(t):D"))
  mod_dd <- lm(formule_dd, data = pseudo_panel)
  mod_dd_rob <- lmtest::coeftest(
    mod_dd,
    vcov = sandwich::vcovCL(mod_dd, cluster = ~grappe)
  )
  cat(sprintf("\n--- DD %s ---\n", outcome))
  print(mod_dd_rob)
}

# ── PSM-DD (Heckman et al. 1997/1998) ────────────────────────
# ATT_PSM-DD = (1/nT) * sum_{i in T} [DeltaY_i - sum_j w_ij DeltaY_j]

cat("\n=== PSM-DD (Heckman 1997/1998) ===\n")
for (outcome in c("pauvre_AF", "pauvre_MODA")) {
  formule_dd <- as.formula(paste(outcome, "~ factor(t) + D + factor(t):D"))
  mod_psm_dd <- lm(formule_dd, data = panel_apparie,
                   weights = panel_apparie$weights)
  mod_psm_dd_rob <- lmtest::coeftest(
    mod_psm_dd,
    vcov = sandwich::vcovCL(mod_psm_dd, cluster = ~grappe)
  )
  cat(sprintf("\n--- PSM-DD %s ---\n", outcome))
  print(mod_psm_dd_rob)
}

# ── Heterogeneite ─────────────────────────────────────────────

cat("\n=== Heterogeneite par milieu ===\n")
for (mil in c(1, 2)) {
  panel_mil <- panel_apparie |> dplyr::filter(f_milieu == mil)
  label_mil <- if (mil == 1) "Urbain" else "Rural"
  for (outcome in c("pauvre_AF", "pauvre_MODA")) {
    formule_dd <- as.formula(paste(outcome, "~ factor(t) + D + factor(t):D"))
    if (nrow(panel_mil) > 30) {
      mod <- lm(formule_dd, data = panel_mil)
      mod_rob <- lmtest::coeftest(mod,
        vcov = sandwich::vcovCL(mod, cluster = ~grappe))
      cat(sprintf("\n--- %s — %s ---\n", label_mil, outcome))
      print(mod_rob)
    }
  }
}

# ── Test tendances paralleles (pre-trend placebo) ─────────────
# Verifier que les tendances etaient similaires avant 2018 (si donnees disponibles)
# ou tester sur une sous-periode interne
cat("\n[Note] Test de tendances paralleles : necessite donnees pre-2018 ou",
    "regression pre-trend sur covariables.\n")
cat("Effectuer test de Rosenbaum pour sensibilite au biais cache (annexe A).\n")

# ── Robustesse : bootstrap (N_BOOT replications) ─────────────

set.seed(SEED)
psm_dd_boot <- function(data, outcome, B = N_BOOT) {
  n <- nrow(data)
  att_boot <- numeric(B)
  for (b in seq_len(B)) {
    idx  <- sample(seq_len(n), n, replace = TRUE)
    boot <- data[idx, ]
    f    <- as.formula(paste(outcome, "~ factor(t) + D + factor(t):D"))
    m    <- try(lm(f, data = boot, weights = boot$weights), silent = TRUE)
    if (!inherits(m, "try-error")) {
      coef_name <- grep("factor\\(t\\)1:D", names(coef(m)), value = TRUE)
      att_boot[b] <- if (length(coef_name)) coef(m)[coef_name] else NA_real_
    }
  }
  att_boot <- att_boot[!is.na(att_boot)]
  list(mean = mean(att_boot), se = sd(att_boot),
       ci95 = quantile(att_boot, c(0.025, 0.975)))
}

cat("\n=== Bootstrap PSM-DD (", N_BOOT, "replications) ===\n")
for (outcome in c("pauvre_AF", "pauvre_MODA")) {
  res <- psm_dd_boot(panel_apparie, outcome)
  cat(sprintf("  %s : ATT=%.4f  SE=%.4f  IC95=[%.4f, %.4f]\n",
              outcome, res$mean, res$se, res$ci95[1], res$ci95[2]))
}

File: code/python/04_psm_dd.py
# ============================================================
#  04_psm_dd.py — Estimation PSM-DD (Heckman et al. 1997/1998)
#  ATT sur pauvre_AF et pauvre_MODA — panel vrai (PanelHH=1)
# ============================================================

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from scipy import stats
from config import BASE_2018, BASE_2021, ID, SEED, OUTPUT_DIR, N_BOOT, CALIPER, K_VOISINS
from utils import lire_stata, taux

np.random.seed(SEED)

# ── 1. Charger le panel vrai ──────────────────────────────────

panel = pd.read_parquet(OUTPUT_DIR / "panel_ind_vrai.parquet")
panel_complet = pd.read_parquet(OUTPUT_DIR / "panel_ind_complet.parquet")

# Covariables welfare
wel_2018_df, _ = lire_stata(BASE_2018, "ehcvm_welfare_sen2018.dta")
wel_2021_df, _ = lire_stata(BASE_2021, "ehcvm_welfare_sen2021.dta")

VARS_WEL = ID + ["pcexp", "hhsize", "region", "milieu",
                 "hgender", "hage", "heduc", "hmstat"]


def prep_welfare(wel):
    cols = [c for c in VARS_WEL if c in wel.columns]
    df = wel[cols].copy()
    df["hhid"] = df["grappe"].astype(int) * 1000 + df["menage"].astype(int)
    return df


def enrichir_welfare(base, wel):
    cols_a_ajouter = [c for c in VARS_WEL + ["hhid"]
                      if c in wel.columns and c not in base.columns and c != "hhid"]
    return base.merge(wel[["hhid"] + cols_a_ajouter], on="hhid", how="left")


wel_18 = prep_welfare(wel_2018_df)
wel_21 = prep_welfare(wel_2021_df)

panel = pd.concat([
    enrichir_welfare(panel[panel["annee"] == 2018].copy(), wel_18),
    enrichir_welfare(panel[panel["annee"] == 2021].copy(), wel_21),
], ignore_index=True)
panel["log_pcexp"] = np.log(panel["pcexp"].clip(lower=1))

panel_complet = pd.concat([
    enrichir_welfare(panel_complet[panel_complet["annee"] == 2018].copy(), wel_18),
    enrichir_welfare(panel_complet[panel_complet["annee"] == 2021].copy(), wel_21),
], ignore_index=True)
panel_complet["log_pcexp"] = np.log(panel_complet["pcexp"].clip(lower=1))

# ── Filtre never-treated ──────────────────────────────────────
# Groupe témoin = D=0 en 2018 ET en 2021 (jamais traité)
# Groupe traité = D=1 en 2018 (statut stable à la période de base)
def filtrer_never_treated(df):
    d_pivot = df.pivot_table(index="hhid", columns="annee", values="D", aggfunc="max")
    if 2018 in d_pivot.columns and 2021 in d_pivot.columns:
        # Traités : D=1 en 2018 ET D=0 en 2021
        traites     = d_pivot[(d_pivot[2018] == 1) & (d_pivot[2021] == 0)].index
        # Témoins : D=0 en 2018 ET D=0 en 2021
        never       = d_pivot[(d_pivot[2018] == 0) & (d_pivot[2021] == 0)].index
        ids_valides = traites.union(never)
    else:
        ids_valides = d_pivot.index
    return df[df["hhid"].isin(ids_valides)].copy()

panel          = filtrer_never_treated(panel)
panel_complet  = filtrer_never_treated(panel_complet)

# D doit être statique (valeur 2018) pour que l'interaction t×D soit valide
def fixer_d_base(df):
    d_base = (df[df["annee"] == 2018][["hhid", "D"]]
              .drop_duplicates("hhid")
              .rename(columns={"D": "D_base"}))
    df = df.drop(columns=["D"]).merge(d_base, on="hhid", how="left")
    df = df.rename(columns={"D_base": "D"})
    return df

panel         = fixer_d_base(panel)
panel_complet = fixer_d_base(panel_complet)

print(f"Panel vrai : {len(panel):,} obs | traites : {panel['D'].sum():,}")

# ── 2. Statistiques descriptives ──────────────────────────────

print("\nStats descriptives par statut de traitement :")
print(panel.groupby(["annee", "D"])[
    ["pauvre_AF", "pauvre_MODA", "score_dep", "pcexp"]
].mean().round(3))

# ── 3. PSM : probit sur t=0 (base 2018) ──────────────────────

base_t0 = panel[panel["annee"] == 2018].dropna(
    subset=["D", "hhsize", "log_pcexp"]
).copy()

# Encodage dummies pour region, milieu, heduc, hmstat
cols_dummies = [c for c in ["region", "milieu", "heduc", "hmstat"]
                if c in base_t0.columns]
base_t0 = pd.get_dummies(base_t0, columns=cols_dummies, drop_first=True)
dummy_cols = [c for c in base_t0.columns
              if any(c.startswith(p + "_") for p in cols_dummies)]

formule = ("D ~ hhsize + log_pcexp + hgender + hage + "
           + " + ".join(dummy_cols)) if dummy_cols else \
          "D ~ hhsize + log_pcexp + hgender + hage"

probit = smf.probit(formule, data=base_t0).fit(disp=False)
print(f"\nPseudo-R2 McFadden : {1 - probit.llf/probit.llnull:.3f}")

base_t0 = base_t0.copy()
base_t0["pscore"] = probit.predict()

# Graphique support commun
fig, ax = plt.subplots(figsize=(8, 5))
for d, label, color in [(0, "Non-traites", "steelblue"), (1, "Traites", "tomato")]:
    base_t0[base_t0["D"] == d]["pscore"].plot.kde(
        ax=ax, label=label, color=color, linewidth=2)
ax.set(title="Distribution du score de propension — panel vrai",
       xlabel="Score de propension", ylabel="Densité")
ax.legend()
fig.tight_layout()
fig.savefig(OUTPUT_DIR / "overlap_panel.pdf")
print("Graphique sauvegardé : overlap_panel.pdf")

# ── 4. Appariement k-NN (k=K_VOISINS, caliper=CALIPER) ───────

from sklearn.neighbors import NearestNeighbors


def apparier_knn(base_t0, k=K_VOISINS, caliper=CALIPER):
    traites = base_t0[base_t0["D"] == 1][["hhid", "pscore"]].reset_index(drop=True)
    temoins = base_t0[base_t0["D"] == 0][["hhid", "pscore"]].reset_index(drop=True)

    nn = NearestNeighbors(n_neighbors=k, algorithm="ball_tree")
    nn.fit(temoins[["pscore"]])
    distances, indices = nn.kneighbors(traites[["pscore"]])

    paires = []
    for i, (dists, idxs) in enumerate(zip(distances, indices)):
        for d, j in zip(dists, idxs):
            if d <= caliper:
                paires.append({
                    "hhid_traite": traites.loc[i, "hhid"],
                    "hhid_temoin": temoins.loc[j, "hhid"],
                    "dist": d,
                })
    return pd.DataFrame(paires)


paires = apparier_knn(base_t0)
ids_traites = paires["hhid_traite"].unique()
ids_temoins = paires["hhid_temoin"].unique()
ids_apparies = np.union1d(ids_traites, ids_temoins)

print(f"\nAppariement k-NN (k={K_VOISINS}, caliper={CALIPER}) :")
print(f"  Traites retenus : {len(ids_traites)}")
print(f"  Temoins uniques : {len(ids_temoins)}")

# SMD avant/apres appariement
cov_cols = ["hhsize", "log_pcexp", "hgender", "hage"]


def smd(x1, x0):
    return (x1.mean() - x0.mean()) / np.sqrt((x1.var() + x0.var()) / 2)


print("\nBalance des covariables (SMD) :")
for col in cov_cols:
    if col in base_t0.columns:
        avant = smd(base_t0[base_t0["D"] == 1][col],
                    base_t0[base_t0["D"] == 0][col])
        apres_data = panel[
            panel["hhid"].isin(ids_apparies) & (panel["annee"] == 2018)
        ].copy()
        if "log_pcexp" not in apres_data.columns:
            apres_data = enrichir_welfare(apres_data, wel_18)
            apres_data["log_pcexp"] = np.log(apres_data["pcexp"].clip(lower=1))
        if col in apres_data.columns:
            apres = smd(apres_data[apres_data["D"] == 1][col],
                        apres_data[apres_data["D"] == 0][col])
            print(f"  {col:20s} : avant={avant:+.3f}  apres={apres:+.3f}")

# ── 5. Double Difference brute (sans appariement, panel vrai) ─

print("\n=== Double Difference brute — panel vrai (sans appariement) ===")
for outcome in ["pauvre_AF", "pauvre_MODA"]:
    mod = smf.ols(f"{outcome} ~ t * D", data=panel).fit(
        cov_type="cluster", cov_kwds={"groups": panel["grappe"]}
    )
    att = mod.params.get("t:D", float("nan"))
    pval = mod.pvalues.get("t:D", float("nan"))
    print(f"  DD {outcome} : ATT = {att:.4f}  (p = {pval:.3f})")

# ── 6. PSM-DD (Heckman 1997/1998) ────────────────────────────

panel_apparie = panel[panel["hhid"].isin(ids_apparies)].copy()
print(f"\nPanel apparie : {len(panel_apparie):,} obs")

print("\n=== PSM-DD — ATT principal (panel vrai) ===")
for outcome in ["pauvre_AF", "pauvre_MODA"]:
    mod = smf.ols(f"{outcome} ~ t * D", data=panel_apparie).fit(
        cov_type="cluster", cov_kwds={"groups": panel_apparie["grappe"]}
    )
    att = mod.params.get("t:D", float("nan"))
    se = mod.bse.get("t:D", float("nan"))
    pval = mod.pvalues.get("t:D", float("nan"))
    ic95_lo = att - 1.96 * se
    ic95_hi = att + 1.96 * se
    print(f"  PSM-DD {outcome} : ATT={att:.4f}  SE={se:.4f}  "
          f"IC95=[{ic95_lo:.4f},{ic95_hi:.4f}]  p={pval:.3f}")

# ── 7. Hétérogénéité par milieu (urbain vs rural) ─────────────

print("\n=== Hétérogénéité par milieu (urbain vs rural) ===")

# Test d'égalité via interaction
if "milieu" in panel_apparie.columns:
    for outcome in ["pauvre_AF", "pauvre_MODA"]:
        mod_int = smf.ols(f"{outcome} ~ t * D * C(milieu)", data=panel_apparie).fit(
            cov_type="cluster", cov_kwds={"groups": panel_apparie["grappe"]}
        )
        print(f"\n  Interaction milieu — {outcome} :")
        coefs_interaction = [(k, v) for k, v in mod_int.params.items()
                             if "milieu" in k and ("t:" in k or ":t" in k)]
        for k, v in coefs_interaction:
            pv = mod_int.pvalues[k]
            print(f"    {k} : coef={v:.4f}  p={pv:.3f}")

for mil, label in [(1, "Urbain"), (2, "Rural")]:
    if "milieu" not in panel_apparie.columns:
        break
    sub = panel_apparie[panel_apparie["milieu"] == mil]
    if len(sub) > 30:
        for outcome in ["pauvre_AF", "pauvre_MODA"]:
            mod = smf.ols(f"{outcome} ~ t * D", data=sub).fit(
                cov_type="cluster", cov_kwds={"groups": sub["grappe"]}
            )
            att = mod.params.get("t:D", float("nan"))
            pval = mod.pvalues.get("t:D", float("nan"))
            print(f"  {label} — {outcome} : ATT={att:.4f}  p={pval:.3f}")

# ── 8. Bootstrap PSM-DD ───────────────────────────────────────

def bootstrap_att(data, outcome, B=N_BOOT, seed=SEED):
    rng = np.random.default_rng(seed)
    att_boot = []
    for _ in range(B):
        idx = rng.integers(0, len(data), len(data))
        boot = data.iloc[idx]
        try:
            m = smf.ols(f"{outcome} ~ t * D", data=boot).fit(disp=False)
            att_boot.append(m.params.get("t:D", np.nan))
        except Exception:
            pass
    arr = np.array([x for x in att_boot if not np.isnan(x)])
    return {
        "mean": arr.mean(),
        "se": arr.std(),
        "ci95": np.percentile(arr, [2.5, 97.5]),
    }


print(f"\n=== Bootstrap PSM-DD ({N_BOOT} replications) — panel vrai ===")
for outcome in ["pauvre_AF", "pauvre_MODA"]:
    res = bootstrap_att(panel_apparie, outcome)
    print(f"  {outcome} : ATT={res['mean']:.4f}  SE={res['se']:.4f}  "
          f"IC95=[{res['ci95'][0]:.4f}, {res['ci95'][1]:.4f}]")

# ── 9. Robustesse : panel vrai vs panel complet ───────────────

print("\n=== Robustesse : panel vrai vs panel complet ===")
for outcome in ["pauvre_AF", "pauvre_MODA"]:
    mod_v = smf.ols(f"{outcome} ~ t * D", data=panel).fit(
        cov_type="cluster", cov_kwds={"groups": panel["grappe"]}
    )
    att_v = mod_v.params.get("t:D", float("nan"))

    mod_c = smf.ols(f"{outcome} ~ t * D", data=panel_complet).fit(
        cov_type="cluster", cov_kwds={"groups": panel_complet["grappe"]}
    )
    att_c = mod_c.params.get("t:D", float("nan"))

    print(f"  {outcome} : panel_vrai={att_v:.4f}  panel_complet={att_c:.4f}")
